src/Versions.jl

```diff
@@ -8,34 +8,37 @@ export VersionBound, VersionRange, VersionSpec, semver_spec, isjoinable
 # VersionBound #
 ################
 struct VersionBound
-    t::NTuple{3,UInt32}
+    t::NTuple{4,UInt32}
     n::Int
     function VersionBound(tin::NTuple{n,Integer}) where n
-        n <= 3 || throw(ArgumentError("VersionBound: you can only specify major, minor and patch versions"))
-        n == 0 && return new((0,           0,      0), n)
-        n == 1 && return new((tin[1],      0,      0), n)
-        n == 2 && return new((tin[1], tin[2],      0), n)
-        n == 3 && return new((tin[1], tin[2], tin[3]), n)
+        n <= 4 || throw(ArgumentError("VersionBound: you can only specify major, minor, patch versions, and a build number"))
+        n == 0 && return new((0,           0,      0,      0), n)
+        n == 1 && return new((tin[1],      0,      0,      0), n)
+        n == 2 && return new((tin[1], tin[2],      0,      0), n)
+        n == 3 && return new((tin[1], tin[2], tin[3],      0), n)
+        n == 4 && return new((tin[1], tin[2], tin[3], tin[4]), n)
         error("invalid $n")
     end
 end
 VersionBound(t::Integer...) = VersionBound(t)
-VersionBound(v::VersionNumber) = VersionBound(v.major, v.minor, v.patch)
+VersionBound(v::VersionNumber) = VersionBound(v.major, v.minor, v.patch, v.build)
```

Comment:
Is this correct? Won't `v.build` be a tuple (not an integer)?

Reply:
true

Reply:
Also, what about the case where no build number is passed, and thus `v.build` is the empty tuple?

Reply:
I guess build should be a `Union{Nothing,Int}` to handle that

Reply:
Well, `VersionBound` currently uses a `t::NTuple{4,UInt32}`. It would be annoying to try to stick a `nothing` into that. How about this?

```suggestion
function VersionBound(v::VersionNumber)
    if isempty(v.build)
        return VersionBound(v.major, v.minor, v.patch)
    else
        return VersionBound(v.major, v.minor, v.patch, only(v.build))
    end
end
```

Reply:
I'm not sure that avoids the compatability with `t::NTuple{4,UInt32}`

Also, instead of `only` it should detect `length > 1` and give a friendly error, I think

Reply:
Perhaps make VersionBound parametric `t::NTuple{V,UInt32}` where V is 3 or 4

Reply:
Isn't that basically already what `VersionBound` does? They have a `t::NTuple{4,UInt32}`, but the "true length" is stored in the `n::Int` field. So if e.g. `n == 2`, then Pkg knows that `t[1]` and `t[2]` have meaningful data, but that `t[3]` should be ignored.